File: Lexer/Nim/lexer.nim
import std/strutils
import token
import std/tables

type Lexer* = object
    input*: string
    char*: char
    curr*: int
    peek*: int

proc readCharacter*(self: var Lexer) =
    if self.peek >= self.input.len:
        self.char = '\0'
    else:
        self.char = self.input[self.peek]
    self.curr = self.peek
    self.peek += 1

proc newLexer*(input: string): Lexer =
    var lex: Lexer = Lexer(input: input, char: '\0', curr: 0, peek: 0)
    lex.readCharacter()
    return lex

proc readIdentifier(self: var Lexer): string =
    let pos = self.curr
    while isAlphaAscii(self.char) or self.char == '_':
        self.readCharacter()
    return self.input[pos..self.curr - 1]

proc readNumber(self: var Lexer): string =
    let pos = self.curr
    while isDigit(self.char):
        self.readCharacter()
    return self.input[pos..self.curr - 1]

proc nextToken*(self: var Lexer): Token =
    while isSpaceAscii(self.char):
        self.readCharacter()

    var tok: Token
    case self.char:
    of '\0': tok = newToken(TokenType.EOF, "\0")
    of '*': tok = newToken(TokenType.ASTERISK, "*")
    of '/': tok = newToken(TokenType.SLASH, "/")
    of '-': tok = newToken(TokenType.DASH, "-")
    of '+': tok = newToken(TokenType.PLUS, "+")
    of '=': tok = newToken(TokenType.EQUAL, "=")
    of '<': tok = newToken(TokenType.LESSTHAN, "<")
    of '>': tok = newToken(TokenType.GREATERTHAN, ">")
    of ';': tok = newToken(TokenType.SEMICOLON, ";")
    of ':': tok = newToken(TokenType.COLON, ":")
    of '(': tok = newToken(TokenType.LPAREN, "(")
    of ')': tok = newToken(TokenType.RPAREN, ")")
    of '{': tok = newToken(TokenType.LBRACE, "{")
    of '}': tok = newToken(TokenType.RBRACE, "}")
    else:
        if isAlphaAscii(self.char):
            let ident = self.readIdentifier()
            if IDENTIFIER_MAP.hasKey(ident):
                return newToken(IDENTIFIER_MAP[ident], ident)
            else:
                return newToken(TokenType.IDENTIFIER, ident)
        elif isDigit(self.char):
            return newToken(TokenType.Number, self.readNumber())
        return newToken(TokenType.Illegal, "ILLEGAL")

    self.readCharacter()
    return tok

File: Lexer/Rust/src/lexer.rs
#![allow(dead_code)]

use std::cell::Cell;

use crate::token;

#[derive(Debug)]
pub struct Lexer<'a> {
    input: &'a str,
    curr: Cell<usize>,
    peek: Cell<usize>,
    char: Cell<char>,
}

impl Lexer<'_> {
    pub fn new(input: &str) -> Lexer {
        let lex = Lexer {
            input,
            curr: Default::default(),
            peek: Default::default(),
            char: Default::default(),
        };
        lex.read_char();
        lex
    }

    pub fn next_token(&self) -> token::Token<'_> {
        while self.char.get().is_whitespace() {
            self.read_char();
        }

        use token::{get_identifier_map, Token, TokenType::*};
        let tok = match self.char.get() {
            '\0' => Token::new(Eof, "\0"),
            '=' => Token::new(Equal, "="),
            '-' => Token::new(Dash, "-"),
            '+' => Token::new(Plus, "+"),
            '(' => Token::new(LParen, "("),
            ')' => Token::new(RParen, ")"),
            '{' => Token::new(LBrace, "{"),
            '}' => Token::new(RBrace, "}"),
            ';' => Token::new(Semicolon, ";"),
            ':' => Token::new(Colon, ":"),
            '<' => Token::new(LessThan, "<"),
            '>' => Token::new(GreaterThan, ">"),
            'a'..='z' | 'A'..='Z' => {
                return {
                    let ident = self.read_identifier();
                    match get_identifier_map().get(ident) {
                        Some(tok_type) => Token::new(*tok_type, ident),
                        None => Token::new(Identifier, ident),
                    }
                }
            },
            '0'..='9' => return Token::new(Number, self.read_number()),
            _ => Token::new(Illegal, "ILLEGAL"),
        };
        self.read_char();
        tok
    }

    fn read_number(&self) -> &str {
        let pos = self.curr.get();
        while self.char.get().is_numeric() {
            self.read_char();
        }
        &self.input[pos..self.curr.get()]
    }

    fn read_identifier(&self) -> &str {
        let pos = self.curr.get();
        while self.char.get().is_alphabetic() || self.char.get() == '_' {
            self.read_char();
        }
        &self.input[pos..self.curr.get()]
    }

    fn read_char(&self) {
        let peek = self.peek.get();
        if peek >= self.input.len() {
            self.char.set('\0')
        } else {
            self.char.set(self.input.as_bytes()[peek].into())
        }
        self.curr.set(peek);
        self.peek.set(peek + 1);
    }
}
